src/main/java/org/rocstreaming/roctoolkit/NativeObject.java

```diff
@@ -1,18 +1,101 @@
 package org.rocstreaming.roctoolkit;
 
-class NativeObject {
+import static java.util.concurrent.TimeUnit.SECONDS;
 
-    private long ptr;
+/**
+ * A <code>NativeObject</code> represents an underlying native roc object.
+ */
+class NativeObject implements AutoCloseable {
 
-    NativeObject() {
-        ptr = 0L;
+    /**
+     * Maximum time to wait (in milliseconds) for joining {@link AutoCloseThread}.
+     */
+    private final static long MAX_JOIN_TIMEOUT_MS = SECONDS.toMillis(20L);
+
+    /**
+     * <code>NativeObject</code> finalizer thread.
+     */
+    private final static AutoCloseThread thread = AutoCloseThread.getInstance();
+
+    /**
+     *  Underlying roc object native pointer.
+     */
+    private final long ptr;
+
+    /**
+     *  Destructor method.
+     */
+    private final Destructor destructor;
```

Comment:
Duplicating fields in NativeObject and NativeObjectReference looks a bit confusing to me.

Currently the flow is the following:

* user creates object derived from NativeObject
* NativeObject initializes its fields (ptr and destructor)
* NativeObject calls AutoCloseThread
* AutoCloseThread creates NativeObjectReference
* NativeObjectReference requests fields from NativeObject and copies them
* AutoCloseThread returns NativeObjectReference
* NativeObject saves NativeObjectReference
* NativeObject calls NativeObjectReference.close(), which uses those saved fields

It was quite hard to me to investigate this sequence.

What do you think about the following change?

* ptr and destructor fields will present only in NativeObjectReference
* NativeObject constructor will create NativeObjectReference (by itself, without help of AutoCloseThread) and pass ptr and destructor to its constructor
* then NativeObject constructor will register created NativeObjectReference to AutoCloseThread
* NativeObject.getPtr() will call NativeObjectReference.getPtr()

I think this way the relationship between NativeObject and NativeObjectReference will be more clear. NativeObjectReference will hold the raw pointer, its destructor, and will implement closing. NativeObject will be a stateless wrapper on top of it, which just contains a reference to NativeObjectReference and registers and de-registers it in AutoCloseThread.

Reply:
Also, maybe rename NativeObject to something else, to avoid confusion with NativeObjectReference? NativeObjectReference looks like it's a reference to NativeObject, but reality is actually the opposite. Maybe rename NativeObject to something like BridgedObject?

Reply:
Actually the `NativeObjectReference` needs a `ReferenceQueue` to be passed to its constructor and it's responsibility of the `AutoCloseThread` to manage the `ReferenceQueue`; so in my opinion it's better to keep `AutoCloseThread` to create a `NativeObjectReference` for better loose coupling.
Surely we can remove the duplicate `ptr` and `destructor` fields from `NativeObject` and let only `NativeObjectReference` to store them.

The new flow will be:
- `NativeObject` constructor calls `AutoCloseThread.add` passing `ptr` and `destructor` to it
- `AutoCloseThread` creates the `NativeObjectReference` (which stores the `ptr` and the `destructor`) and returns it
- `NativeObject` saves `NativeObjectReference`
- `NativeObject.getPtr()` will call `NativeObjectReference.getPtr()`

Reply:
> Also, maybe rename NativeObject to something else, to avoid confusion with NativeObjectReference? NativeObjectReference looks like it's a reference to NativeObject, but reality is actually the opposite. Maybe rename NativeObject to something like BridgedObject?

Sure I'll rename it to `BridgedObject`

Reply:
Sounds good!